Hw05 parallel execution

ZaytsveDmitriy opened the pull request:
## Домашнее задание №5 «Параллельное исполнение»

### Чек-лист студента ([Что это?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0))
- [x] Я перечитал задание после решения. [Зачем перечитывать задание?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%82%D1%8B%D0%B2%D0%B0%D1%82%D1%8C-%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5)
- [x] Я запустил `go mod tidy`.
- [x] Я удалил `.sync` файл. [Зачем его удалять?](https://github.com/OtusGolang/home_work/wiki/%5B%D0%A1%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0%D0%BC%5D-%D0%9F%D1%80%D0%BE%D1%86%D0%B5%D1%81%D1%81-%D1%81%D0%B4%D0%B0%D1%87%D0%B8-%D0%94%D0%97#user-content-%D0%92%D0%B0%D1%80%D0%B8%D0%B0%D0%BD%D1%82-2)
- [x] Я сравнил объем кода с объемом авторского решения. [Зачем сверять объем кода?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D1%81%D0%B2%D0%B5%D1%80%D1%8F%D1%82%D1%8C-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BC-%D0%BA%D0%BE%D0%B4%D0%B0)
  - run.go ~55 строк

### Критерии оценки
- [ ] Пайплайн зелёный - 4 балла
- [ ] Добавлены новые юнит-тесты - до 4 баллов
- [ ] Понятность и чистота кода - до 2 баллов

#### Зачёт от 7 баллов

ZaytsveDmitriy committed:
add implementation and tests

ZaytsveDmitriy committed:
delete .sync

dshemin commented on hw05_parallel_execution/run.go:
```diff
@@ -2,14 +2,72 @@ package hw05parallelexecution
-// Run starts tasks in n goroutines and stops its work when receiving m errors from tasks.
 func Run(tasks []Task, n, m int) error {
-	// Place your code here.
+	if len(tasks) < 1 {
```
Важно:

Нет проверки на значение `n`. Надо проверить и добавить тест.

dshemin commented on hw05_parallel_execution/run.go:
```diff
@@ -2,14 +2,72 @@ package hw05parallelexecution
 	return nil
 }
+
+func worker(chTasks <-chan Task, errCnt *atomic.Int64, chClose <-chan struct{}, wg *sync.WaitGroup) {
```
Важно:

Не добавляйте sync.WaitGroup как аргумент в функцию - это создаёт дополнительную и не нужную зависимость которая не влияет на логику работы функции, усложняет переиспользование и тестрование.

ZaytsveDmitriy committed:
Edit func Run and worker, add new test

- func Run: add handle for n worker less 1
- func worker: exclude WhaitGroup from func in args
- add test: tasks with zerro allow workers
